Add BleBox integration

gadgetmobile opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change

This is a WIP for a complete/quality-oriented integration for BleBox smart home devices.


### Features:
  - 11 [BleBox products](https://blebox.eu/?lang=en) are implemented. **(PR reduced to 1 platform/3 devices)**
  - almost 100% coverage through tests (which mock out 100% of the client PyPI lib)
  - minimal, bare-bones implemenation on Hass side (code is almost 100% Hass-specific)
  - separate PyPI client library doing all the work (with its end-to-end functional tests)


NOTE: I first wanted some feedback for the overall approach and testing
strategy - so pick a single platform yet (as a first PR) didn't seem to make
sense. (Let me know if I'm wrong and I can reduce this PR to one platform).

I'd also really appreciate some tips and feedback on:
  - overall do's/don'ts in Hass integrations (that I got wrong or missed)
  - what other major changes are needed to make this a decent PR?
  - how I can make reviewing easier

Also, thank you all for putting so much care and time into making Home
Assistant such a rich and high quality automation platform!


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13148


### Issues / questions

How should I properly handle ...

- [X]  ... `async_update` - when multiple entities use same device (throttling? via `@Throttle`?)
- [X]  ... configuration (via config_flow and otherwise) - best practices ?
- [X]  ... device identification - async ? (instead of synchronously during setup)
- [X]  ... persistently storing state (e.g. "last on color/brightness")?
- [X]  ... misc sensors - "as many as possible" or conservatively (only when requested by users)?
- [X]  ... exceptions - are harmless call stacks ok in logs? or should I always catch+log errors?
- [X]  ... serializing requests to devices (due to device limitations)

(Most of the remaining things I was unsure about are marked as TODO's in the code).

Any more tips/guidance to make this a great integration are appreciated.


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

gadgetmobile committed:
add BleBox integration (11 devices)

homeassistant commented:
Hi @gadgetmobile, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#32664).

Once you do that we will be able to review and accept this pull request.

Thanks!

gadgetmobile committed:
bump blebox_uniapi version

Bump to fix switchBox newer API support

gadgetmobile commented:
NOTE: Build failures are only due to pylint returning '4' (due to harmless TODO's):

https://dev.azure.com/home-assistant/Core/_build/results?buildId=31768&view=logs&j=8b66295a-05e1-52cf-5640-7a4f4e1975e2&t=6c4465fe-285a-5650-787c-2e3dbe1da192&l=48

(Most TODO's mark missing coverage points, the remainder are improvement suggestions).

gadgetmobile committed:
disable harmless pylint warnings (TODOs)

frenck commented:
Please reduce the PR to a single platform. These kinds of large PR's are hard to review. As per: <https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible>

As for the TODO's; In general, it is better to make it trackable, having it in code often ends up being there forever. The rules are enabled for that reason. If you really have the intention to fix those; create a tracking issue after the PR is merged.

gadgetmobile committed:
increase coverage + cleanup TODOs

gadgetmobile committed:
limit PR to one platform (covers)

gadgetmobile committed:
increase coverage + cleanup TODOs

gadgetmobile committed:
remove unused pylint comment

gadgetmobile committed:
set PARALLEL_UPDATES=1 for current blebox_uniapi

gadgetmobile commented:
> Please reduce the PR to a single platform.

Done. (Only covers + config flow).

> As for the TODO's (...)

I planned to deal with all of them before squashing the commits. I've removed all of them. (Along with pylint comments disabling them). Coverage is also now at 100%. (Though I wish there were standardized ways to test things like `async_setup`, etc.).

> https://developers.home-assistant.io/docs/integration_fetching_data/#request-parallelism

Thanks! I must have missed it. In my case, I understand it's "0" (since every entity defines `async_update`) and it's sufficient for me to use a semaphore in the client library.


> > ... configuration (via config_flow and otherwise) - best practices ?
> 
> Not sure about your question here?

All I really need to configure is the IP address (and port, though it's always 80). I'm not sure what other cases to handle, if any. (Like preventing adding and already existing device/ip - which I'm doing already).

> If a connection is needed for device identification, that is possible (Hue & WLED do this as well).

I didn't know WLED was a good example, but I seem to be doing exactly the same thing - except for handling the mac address (with `async_set_unique_id` and `_abort_if_unique_id_configured`). I'll change that later.

> In general, we try to make the most usable data available in separate entities. However, we do recommend to disable less common ones by default.

Makes sense, thanks. I'll keep that in mind once more sensors are requested.

> Logs should be usable and readable for the average user, stack traces are not part of that. 

Thanks. I'll make sure handle them appropriately (along with proper tests and coverage).

Follow up questions:
1. BleBox devices are feature-rich wireless devices that speak HTTP/JSON. Are there any recommendations or other integrations worth looking at or comparing to (beside HUE and WLED)?
1. Should I add the physical devices to the registry? (Is it useful or clutter?)
1. BleBox devices generally respond to commands (POST) with complete current state - should I do something to skip unnecessary/immediate `async_update` calls when this happens? (I was thinking of just skipping/throttling updates in the client library based on recency of data).
1. I am confused as to when `async_setup_platform` is required to be implemented. Do I need it?

Thanks again for helping me getting this done right! (I really appreciate it!)

gadgetmobile committed:
remove TODO comment

gadgetmobile committed:
add missing errors.py file

gadgetmobile committed:
reorganize strings + translations

gadgetmobile committed:
use a fixed timeout of 3 seconds

gadgetmobile committed:
allow parallel updates

gadgetmobile committed:
rework config flow

gadgetmobile committed:
bump blebox_uniapi to 1.2.0

gadgetmobile committed:
rework error handling + increase coverage

gadgetmobile committed:
increase coverage to 100%

gadgetmobile committed:
fix pylint error

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,133 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add, discovery_info=None):
```
You should be able to remove this since its setup via the config entry and not a platform

gadgetmobile replied:
Removed, thanks!

bdraco commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,148 @@
+    )
+
+
+class DefaultBoxTest:
```
New tests need to be standalone pytest test functions (no classes).

gadgetmobile replied:
As mentioned elsewhere, I need this class for shared tests (used in other platforms, removed from this PR) and I can't figure out how properly implement shared tests.

bdraco commented on tests/components/blebox/test_cover.py:
```diff
@@ -0,0 +1,379 @@
+from .conftest import DefaultBoxTest, mock_feature
+
+
+class CoverTest(DefaultBoxTest):
```
New tests need to be standalone pytest test functions (no classes).

bdraco commented on tests/components/blebox/test_cover.py:
```diff
@@ -0,0 +1,379 @@
+        self.assert_state(entity, STATE_OPEN)
+
+
+class TestShutter(CoverTest):
```
New tests need to be standalone pytest test functions (no classes).

bdraco commented on tests/components/blebox/test_cover.py:
```diff
@@ -0,0 +1,379 @@
+            )
+
+
+class TestGateBox(CoverTest):
```
New tests need to be standalone pytest test functions (no classes).

gadgetmobile replied:
Fixed, along with refactoring using pytest fixture parameters.

However, I reuse some tests and helpers (defined in `DefaultBoxTest`) in other platforms (not included in this PR).

So for now, I'm calling those shared tests directly from `DefaultBoxTest` in case there are more changes needed in `test_cover.py`. (Since I'd have to update other platforms not included in this PR).

I'm not sure how to cleanly implement shared tests without classes, though.

bdraco replied:
Your mocks can be classes, its just the tests themselves that need to be standalone pytest test functions

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+        errors = {}
+        if user_input is not None:
+
+            def host_port(data):
```
This one doesn't need to be inlined, please break it out

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,104 @@
+
+async def async_add_blebox(klass, method, hass, config, async_add, exception):
+    """Add a BleBox device from the given config."""
+    host = config.get(CONF_HOST)
```
```suggestion
    host = config[CONF_HOST]
```
It looks like this is a required value

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,104 @@
+async def async_add_blebox(klass, method, hass, config, async_add, exception):
+    """Add a BleBox device from the given config."""
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
```
```suggestion
    port = config[CONF_PORT]
```
It looks like this is a required value

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,104 @@
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    # timeout = config.get(CONF_TIMEOUT)
+    timeout = 3
```
Please make this a constant in const.py

gadgetmobile replied:
Fixed, thanks!

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,104 @@
+    return True
+
+
+class CommonEntity:
```
```suggestion
class BleBoxEntity:
```

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,133 @@
+    )
+
+
+class BleBoxCoverEntity(CommonEntity, CoverDevice):
```
```suggestion
class BleBoxCoverEntity(BleBoxEntity, CoverDevice):
```

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+            host = user_input[CONF_HOST]
+            port = user_input[CONF_PORT]
+        else:
+            host = PLACEHOLDER_HOST
```
Looks like you used `description_placeholders` above but not here?

gadgetmobile replied:
I refactored config_flow.py to fix this (and above comments) and to make reviewing much easier.

bdraco reviewed (changes requested):
Hi @gadgetmobile   Thanks for submitting this integration!  Please take a look at my review comments.

gadgetmobile committed:
remove async_setup_platform

gadgetmobile committed:
Update homeassistant/components/blebox/__init__.py

require CONF_HOST in config

Co-Authored-By: J. Nick Koston <nick@koston.org>

gadgetmobile committed:
Update homeassistant/components/blebox/__init__.py

require CONF_PORT in config

Co-Authored-By: J. Nick Koston <nick@koston.org>

gadgetmobile committed:
Update homeassistant/components/blebox/__init__.py

rename CommonEntity to BleBoxEntity

Co-Authored-By: J. Nick Koston <nick@koston.org>

gadgetmobile committed:
Update homeassistant/components/blebox/cover.py

rename CommonEntity to BleBoxEntity

Co-Authored-By: J. Nick Koston <nick@koston.org>

gadgetmobile committed:
make timeout a const

gadgetmobile committed:
refactor config_flow

gadgetmobile committed:
rename CommonEntity to BleBoxEntity

gadgetmobile committed:
improve error handling

gadgetmobile committed:
bump to blebox_uniapi 1.3.0

gadgetmobile committed:
add device registry entries

gadgetmobile committed:
fix docsstring

gadgetmobile committed:
raise if set_cover_position is not supported

gadgetmobile committed:
remove test classes

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,112 @@
+
+    entities = []
+    for entity in product.features[method]:
+        entities.append(klass(entity))
```
Its hard to follow what is going on here since there are Home Assistant entities and BleBox entities where are then referred to as `features` in `BleBoxEntity`.  Would you please name the BleBox entities consistantly
```suggestion
        entities.append(klass(product))
```

gadgetmobile replied:
I updated the docstring and variables so I hope it makes a lot more sense now.

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,112 @@
+        product = await Products.async_from_host(api_host)
+    except Error as ex:
+        _LOGGER.error("Identify failed at %s:%d (%s)", host, port, ex)
+        raise exception from ex
```
```suggestion
        raise
```

bdraco replied:
If its a connection error you can raise `ConfigEntryNotReady` to have this retried later

gadgetmobile replied:
I tried `ConfigEntryNotReady` at first, except it doesn't seem to work.

(That's why I added the parameter - so I could pick exceptions once I could figure out which scenario works in practice with which exception).

Instead, using PlatformNotReady works as expected (handled + logging just the error message + HA retries the setup multiple times).

Scenario:
1. I configure a device (user flow).
2. I disconnect the physical device from the network.
3. I restart HA
4. Exception backtrace is logged when ConfigEntryNotReady is used (raises ConfigEntryNoReady, which is never caught and no retries), but works as expected when PlatformNotReady is used (expected = only logs error, with no exception backtrace, then retries).

During HA startup, it seems the setup is handled in `entity_platform.py` which does not handle `ConfigEntryNotReady` exceptions. (I can't even figure out how to trigger a case where ConfigEntryNotReady is likewise handled with retries/etc.)

bdraco replied:
https://developers.home-assistant.io/docs/creating_component_generic_discovery
"Most integrations consist of a single platform. And in that case, it's fine to just define that one platform. However, if you are going to add a second platform, you will want to centralize your connection logic. This is done inside the component (__init__.py)."

https://developers.home-assistant.io/docs/config_entries_index

You'll need to raise ConfigEntryNotReady in `async_setup_entry` in `__init__.py`

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+        """Initialize the BleBox config flow."""
+        self.device_config = {}
+
+    def handle(self, step, exception, schema, addr, message_id):
```
```suggestion
    def handle_step_exception(self, step, exception, schema, addr, message_id):
```

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,129 @@
+        hass,
+        config_entry.data,
+        async_add,
+        PlatformNotReady,
```
```suggestion
```
Please raise `ConfigEntryNotReady` in `__init__` if there is a connection failure instead

gadgetmobile replied:
As above, ConfigEntryNotReady doesn't seem to work in practice.

(I'll refactor this out once the ConfigEntryNotReady/PlatformNotReady issue is figured out).

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,129 @@
+    @property
+    def state(self):
+        """Return the equivalent HA cover state."""
+        states = {
```
```suggestion
        BLEBOX_TO_HASS_STATES = {
```
Please move this to the top or in const.py

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,129 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        types = {
```
```suggestion
        BLEBOX_TO_HASS_DEVICE_CLASSES = {
```
Please move this to the top or in const.py

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,129 @@
+
+    async def async_set_cover_position(self, **kwargs):
+        """Set the cover position."""
+        if not SUPPORT_SET_POSITION & self.supported_features:
```
```suggestion
        if not self.supported_features & SUPPORT_SET_POSITION:
```

bdraco commented on homeassistant/components/blebox/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "name": "BleBox devices",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/blebox",
+  "requirements": ["blebox_uniapi==1.3.0", "semver==2.9.1"],
```
```suggestion
  "requirements": ["blebox_uniapi==1.3.0"],
```
Doesn't look like semver is used

gadgetmobile replied:
Thanks!

bdraco commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,147 @@
+        """Return a default entity mock."""
+        raise NotImplementedError("Implement me")  # pragma: no cover
+
+    async def test_update_failure(self, hass):
```
New tests need to be standalone pytest test functions.  If the function starts with `test_`, please move it out of the class

bdraco commented on tests/components/blebox/test_config_flow.py:
```diff
@@ -0,0 +1,197 @@
+    """Test async_unload_entry (for coverage)."""
+    config = mock_config("172.90.80.70")
+    config.add_to_hass(hass)
+    with patch.object(
```
Please don't patch home assistant internals as they may change.

gadgetmobile replied:
As I understand, `async_forward_entry_setup` is the public+recommended API to use for `async_setup_entry` and `async_unload_entry`.

Since it's public, I need to mock it for coverage testing of those setup/unload methods.

(Unless I misunderstood what you mean).

bdraco commented on tests/components/blebox/test_config_flow.py:
```diff
@@ -0,0 +1,197 @@
+    config.add_to_hass(hass)
+
+    with patch.object(
+        hass.config_entries,
```
Please don't patch home assistant internals as they may change.

bdraco replied:
You could do something like 
```
from tests.common import MockConfigEntry, load_fixture
...
    with patch("some_part_of_blebox_pypi_package", side_effect=_handle_mock_api_request):
        entry = MockConfigEntry(
            domain=DOMAIN, data={CONF_USERNAME: "mock", CONF_PASSWORD: "mock"}
        )
        entry.add_to_hass(hass)
        await hass.config_entries.async_setup(entry.entry_id)
        await hass.async_block_till_done()

  .. check that something you expect to be there is ...
```

gadgetmobile replied:
Fixed in https://github.com/home-assistant/core/pull/32664/commits/0af15592cea1ca8b92f3f89631c34f0711cd0d02

(At first I didn't realize you meant to use higher level tests, which do make a lot more sense).

As for grouping tests together, I think config_entries setup/remove tests shouldn't be in `test_config_flow.py` (which is basically just for user flow in my case). They should probably be repeated for every platform I guess. But that may create a lot more repetitive tests that could slow down CI builds noticeably. If you have suggestions, let me know, thanks!

bdraco commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,129 @@
+
+async def async_setup_entry(hass, config_entry, async_add):
+    """Set up a BleBox entry."""
+    return await async_add_blebox(
```
```suggestion
    return await async_create_blebox_entities(
```

bdraco reviewed (changes requested):
Hi @gadgetmobile Thanks for working through these.  There are a few more things to address.

gadgetmobile committed:
remove unused semver dependency

gadgetmobile committed:
rename handle to handle_step_exception

gadgetmobile commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,129 @@
+        types = {
+            "shutter": DEVICE_CLASS_SHUTTER,
+            "gatebox": DEVICE_CLASS_DOOR,
+            "gate": DEVICE_CLASS_DOOR,
```
No such thing in Home Assistant as DEVICE_CLASS_GATE in the codebase nor git history.

(Only in the docs here: https://github.com/home-assistant/developers.home-assistant/commit/b1ad6ba137447e3a3419394f9d8d9851c245931e)

bdraco replied:
Your branch is a bit behind dev.

https://developers.home-assistant.io/docs/development_catching_up/

gadgetmobile committed:
change param order for bitwise AND operator

gadgetmobile committed:
move state and device class maps to consts

gadgetmobile committed:
rename method to async_create_blebox_entities

gadgetmobile committed:
improve naming clarity and consistency

gadgetmobile committed:
refactor async_create_blebox_entities

gadgetmobile committed:
use new DEVICE_CLASS_GATE constant

gadgetmobile committed:
regenerate dependency requirements

gadgetmobile committed:
update broken test

gadgetmobile committed:
fix pylint error

gadgetmobile committed:
rework shared tests

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,105 @@
+        try:
+            await self._feature.async_update()
+        except Error as ex:
+            _LOGGER.error("Updating '%s' failed: %s", self.name, ex)
```
If the update fails it would be nice to show that its not available anymore
```
   self._available = False
```

```
    @property
    def available(self):
        """...."""
        return self._available
```

bdraco replied:
https://developers.home-assistant.io/docs/integration_quality_scale_index/
`Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.`

The standard only speaks to handle the logging.  If you think it needs 3 update failures to declare unavailable that seems fine to me, however someone might feel differently.

Probably best to leave for now and add later in a followup as to not hold things up.

gadgetmobile replied:
I think it's best for the client library to inform about the availability status after every update.

(Especially since the "algorithm" for it can be device-specific and even case-specific).

Then, the Home Assistant BleBox integration would only handle logging according to the change in availability.

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,105 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
```
Looks like this isn't being used since there is no yaml config.  If you want to allow yaml config you can do something like this
```
async def async_setup(hass: HomeAssistant, config: dict) -> bool:
    """Set up blebox from YAML."""

    hass.data.setdefault(DOMAIN, {})

    if not config.get(DOMAIN):
     return True

    hass.async_create_task(
        hass.config_entries.flow.async_init(
            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
        )
    )
    return True
```

gadgetmobile replied:
Probably the only config option I'd want to is something like a custom SCAN_INTERVAL or related. (Though this probably only makes sense for testing).

If I can handle that from the config flow, I believe I can do without the YAML config.

(BleBox has very decent and feature-rich app/web UIs to configure their devices - so duplicating this would only unnecessarily increase Home Assistant maintainer work).

So I'll remove it tomorrow (with the last batch of fixes for this PR).

Thanks for putting in so much time and care into this review! I really appreciate it!

bdraco replied:
Ideally as developers we choose a good SCAN_INTERVAL and it shouldn't be something the user has to configure as we should know more about what is going on (hopefully)

gadgetmobile committed:
remove unused CONFIG_SCHEMA

gadgetmobile committed:
refactor + improve config_flow tests

gadgetmobile committed:
fix typo in TODO

gadgetmobile committed:
bump blebox_uniapi dependency to 1.3.1

gadgetmobile committed:
use ConfigEntryNotReady and rework setup

gadgetmobile commented:
Changes since last review:

- Rebased to dev branch
- 548930ed6 using ConfigEntryNotReady properly, setup improved (Api client setup moved to __init__).
- 7a48577ce blebox uniapi bump to 1.3.1

I'll double check to make sure I haven't missed anything.

Let me know if there's anything else I can correct/improve, thanks!

gadgetmobile committed:
fix pyupgrade errors

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+    def handle_step_exception(self, step, exception, schema, addr, message_id):
+        """Handle step exceptions."""
+
+        _LOGGER.error("%s at %s:%d (%s)", LOG_MSG[message_id], *addr, exception)
```
```suggestion
        _LOGGER.debug("%s at %s:%d (%s)", LOG_MSG[message_id], *addr, exception)
```

bdraco replied:
If we know what the exception is we shouldn't log it.  Probably ok for unknown.

gadgetmobile replied:
As suggested, except a warning is shown for connection errors. (These can be common and hard to diagnose, while the translated message explains to "see the logs", and these errors are too important - for now - to expect users to enable debug logs).

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+    UNSUPPORTED_VERSION,
+)
+
+DEFAULT_HOST = "192.168.0.2"
```
Could moved these to `const.py` but doesn't really matter

bdraco commented on homeassistant/components/blebox/strings.json:
```diff
@@ -0,0 +1,27 @@
+      "already_configured": "This BleBox device is already configured.",
+      "address_already_configured": "A BleBox device is already configured at {address}.",
+      "no_devices_found": "No BleBox devices devices found on the network.",
+      "single_instance_allowed": "Only a single configuration of BleBox devices is possible."
```
```suggestion
```

Looks unused

bdraco commented on homeassistant/components/blebox/strings.json:
```diff
@@ -0,0 +1,27 @@
+    "abort": {
+      "already_configured": "This BleBox device is already configured.",
+      "address_already_configured": "A BleBox device is already configured at {address}.",
+      "no_devices_found": "No BleBox devices devices found on the network.",
```
```suggestion
```

Looks unused?

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,117 @@
+            ]
+        )
+    )
+
```
We need to delete the data as well.

Usually index by the config entry id

`hass.data[DOMAIN][config_entry.entry_id][PRODUCTS]`
instead of
`hass.data[DOMAIN][PRODUCTS][config_entry.entry_id]`



so we can do this here: 

```suggestion
    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)
```

bdraco commented:
@gadgetmobile Seems like this is getting close to merge, would you please recheck the docs to make sure they are still accurate.

gadgetmobile committed:
log errors only for unknown exceptions

gadgetmobile committed:
move constants to consts.py

gadgetmobile committed:
remove unused translation strings

gadgetmobile committed:
add comment for translations string consts

gadgetmobile committed:
improve component unloading + missing test

gadgetmobile commented:
> @gadgetmobile Seems like this is getting close to merge, would you please recheck the docs to make sure they are still accurate.

Once the review is finished here, I'll be sure have the docs updated (and the PR linked here) before I squash all the above commits.

Thanks for all your help and support so far!

gadgetmobile committed:
fix integration translation validation for v0.109

bdraco commented:
@gadgetmobile No need to squash, we do that on merge

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+
+        log_fn("%s at %s:%d (%s)", LOG_MSG[message_id], *addr, exception)
+
+        address = "{}:{}".format(*addr)
```
The home assistant standard is to use f-strings instead of .format

bdraco commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+    def abort_because_configured(self, addr):
+        """Return abort flow response for when already configured."""
+
+        address = "{}:{}".format(*addr)
```
The home assistant standard is to use f-strings instead of .format

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,120 @@
+
+    @property
+    def device_info(self):
+        """Return device information about this WLED device."""
```
Stale doc string

bdraco commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,120 @@
+    """Implements methods common among BleBox entities."""
+
+    def __init__(self, feature):
+        """Initialize the cover entity."""
```
Stale docstring

gadgetmobile committed:
update docstrings

gadgetmobile committed:
avoid using format()

gadgetmobile commented:
I believe I've addressed everything (including docs and brand images).

Let me know if I've missed anything.

Thanks again so much!

bdraco commented on homeassistant/components/blebox/errors.py:
```diff
@@ -0,0 +1 @@
+"""Errors for the blebox component."""
```
Looks like this can be deleted?

bdraco commented on homeassistant/components/blebox/.translations/en.json:
```diff
@@ -0,0 +1,24 @@
+{
```
Translations have moved to 'translations' instead of '.translations' in the latest version

bdraco commented on tests/components/blebox/test_config_flow.py:
```diff
@@ -0,0 +1,181 @@
+
+@pytest.fixture
+def flow(hass):
+    """Return a user configuration flow."""
```
@MartinHjelmare Is this ok or does he need to call `async_init` instead?

```
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
```

gadgetmobile replied:
Fixed and refactored, thanks!

gadgetmobile committed:
remove unused file

gadgetmobile committed:
rename translations dir to match new HA version

gadgetmobile committed:
properly use async_init in tests + refactor

gadgetmobile commented:
I fixed the above issues with the following and rebased (to match translations dir change):

38581a5fb6 remove unused file (errors.py)
6f606cdaf4 rename translations dir to match new HA version
533578fa46 properly use async_init in tests + refactor

P.S. CI seems to have failed on an unrelated test (Timeout in test_form_invalid_auth in vilfo).

bdraco commented:
@gadgetmobile I fixed the CI here https://github.com/home-assistant/core/pull/34647 once it merges it should run cleanly after rebase

bdraco commented:
@gadgetmobile CI should be ok now after

`git rebase upstream/dev`

gadgetmobile commented:
CI check succeeds now, thanks. (Nothing was changed, except rebase).

MartinHjelmare commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,120 @@
+    return unload_ok
+
+
+async def async_create_blebox_entities(product, async_add, entity_klass, entity_type):
```
This doesn't need to be a coroutine since we don't await inside.

Please decorate the function with `@callback` imported from core.py.

MartinHjelmare replied:
```suggestion
@callback
def async_create_blebox_entities(product, async_add, entity_klass, entity_type):
```

MartinHjelmare commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,120 @@
+        entities.append(entity_klass(feature))
+
+    async_add(entities, True)
+    return True
```
Nothing is checking this return value.
```suggestion
```

MartinHjelmare commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -0,0 +1,120 @@
+    return True
+
+
+class BleBoxEntity:
```
This should inherit `Entity`.
```suggestion
class BleBoxEntity(Entity):
```

gadgetmobile replied:
Already inherited indirectly later (via a different class), but added it for clarity anyway.

MartinHjelmare commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+            description_placeholders={"address": f"{host}:{port}"},
+        )
+
+    def abort_because_configured(self, host, port):
```
This doesn't seem needed. Just inline the body in `async_step_user`.

MartinHjelmare commented on homeassistant/components/blebox/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+                return self.abort_because_configured(*addr)
+
+        websession = async_get_clientsession(hass)
+        api_host = ApiHost(*addr, DEFAULT_SETUP_TIMEOUT, websession, hass.loop, _LOGGER)
```
We should not pass our logger to the library. The library should use its own per module logger as recommended in the standard library:
```py
LOGGER = logging.getLogger(__name__)
```

MartinHjelmare commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,109 @@
+    """Set up a BleBox entry."""
+
+    product = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
+    return await async_create_blebox_entities(
```
Nothing is checking this return value.
```suggestion
    async_create_blebox_entities(
```

MartinHjelmare commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,109 @@
+    )
+
+
+class BleBoxCoverEntity(BleBoxEntity, CoverDevice):
```
We've renamed our entity classes.
```suggestion
class BleBoxCoverEntity(BleBoxEntity, CoverEntity):
```

MartinHjelmare commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,109 @@
+        """Return the current cover position."""
+        position = self._feature.current
+        if position == -1:  # possible for shutterBox
+            name = self.name
+            _LOGGER.warning(
+                "Position for %s is unknown. Try calibrating the device.", name
+            )
```
We don't want side effects in entity state properties. Please remove the logging.
```suggestion
```

MartinHjelmare commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,109 @@
+
+    async def async_set_cover_position(self, **kwargs):
+        """Set the cover position."""
+        if not self.supported_features & SUPPORT_SET_POSITION:
```
It would be better to fix the service registration in the core. The core should guard this service entity method in the core service handler, depending on what supported features the entity supports.

I'll make a PR with this and we can clean up here.

MartinHjelmare replied:
PR here:
https://github.com/home-assistant/core/pull/35073

MartinHjelmare commented on homeassistant/components/blebox/cover.py:
```diff
@@ -0,0 +1,109 @@
+            raise NotImplementedError
+
+        position = kwargs[ATTR_POSITION]
+        if position is not None:
+            await self._feature.async_set_position(100 - position)
```
Position is validated by the service schema.
```suggestion
        await self._feature.async_set_position(100 - position)
```

MartinHjelmare commented on homeassistant/components/blebox/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/blebox",
+  "requirements": ["blebox_uniapi==1.3.1"],
+  "dependencies": [],
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/blebox/const.py:
```diff
@@ -0,0 +1,45 @@
+DEFAULT_SETUP_TIMEOUT = 3
+
+# translation strings
+ALREADY_CONFIGURED = "address_already_configured"
```
```suggestion
ADDRESS_ALREADY_CONFIGURED = "address_already_configured"
```

MartinHjelmare commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,105 @@
+
+from unittest import mock
+
+from asynctest import CoroutineMock, PropertyMock, patch
```
Please use our test mock helper instead. See below.
```suggestion
```

MartinHjelmare commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,105 @@
+from homeassistant.components.blebox import const
+from homeassistant.const import CONF_HOST, CONF_PORT
+
+from tests.common import MockConfigEntry
```
```suggestion
from tests.async_mock import AsyncMock, PropertyMock, patch
from tests.common import MockConfigEntry
```

MartinHjelmare commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,105 @@
+
+    patcher = patch(path, mock.DEFAULT, blebox_uniapi.products.Products, True, True)
+    products_class = patcher.start()
+    products_class.async_from_host = CoroutineMock(**kwargs)
```
```suggestion
    products_class.async_from_host = AsyncMock(**kwargs)
```

MartinHjelmare commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,105 @@
+        config = mock_config()
+        config.add_to_hass(hass)
+
+        domain = hass.data.setdefault(const.DOMAIN, {})
```
We should not mock integration code. Set up the integration instead.

MartinHjelmare commented on tests/components/blebox/conftest.py:
```diff
@@ -0,0 +1,105 @@
+    )
+
+
+class BleBoxTestHelper:
```
Please remove this helper. We should just set up the integration with a mock config entry, and then assert the state of the config entry, and the state of the entities via the core state machine.

We should not mock integration code and we should not interact with integration code directly. This will make the tests robust for future changes.

gadgetmobile replied:
To make things even more robust and cleaner, I transformed the helper into a wrapper class for more convenient entities/state access - and moved it to `conftest.py`.

MartinHjelmare commented on tests/components/blebox/test_cover.py:
```diff
@@ -0,0 +1,446 @@
+
+def assert_state(entity, state):
+    """Assert that cover state is correct."""
+    assert entity.state == state
```
Don't interact with the entity directly. Get the state of the entity from the core state machine and assert the core state. The state is the truth of the entity.

MartinHjelmare replied:
Here's an example:
https://github.com/home-assistant/core/blob/0519b96bec79b2db2ec8021aca1149288e2b659a/tests/components/heos/test_media_player.py#L66-L100

gadgetmobile committed:
remove async property on non-async method

gadgetmobile committed:
remove return value which is never used

gadgetmobile committed:
implicitly inherit Entity for clarity

gadgetmobile committed:
inline method which is used only once

gadgetmobile committed:
bump blebox_uniapi version to 1.3.2

gadgetmobile committed:
stop sharing logger with blebox_uniapi

gadgetmobile committed:
remove property side-effect

gadgetmobile committed:
avoid inheriting deprecated CoverDevice class

gadgetmobile committed:
assume core handles unsupported case properly

gadgetmobile committed:
remove unnecessary position presence checking

gadgetmobile committed:
remove empty dependencies section in manifest

gadgetmobile committed:
rename constant to ADDRESS_ALREADY_CONFIGURED

gadgetmobile committed:
use tests.async_mock instead of asynctest

gadgetmobile committed:
rework cover tests and helper classes

MartinHjelmare commented on tests/components/blebox/conftest.py:
```diff
@@ -59,47 +65,96 @@ def mock_feature(category, spec, **kwargs):
-    """Helper methods for tests."""
 
-    HASS_TYPE = None  # override in subclass
+class Wrapper:
```
It's ok to use a helper for the set up part of the tests but not for the action or assertion. Please at most just make a helper function that sets up the integration.

gadgetmobile replied:
I'm more a fan of clearer intentions in tests (instead of removing the fascade, revealing details with the cost of more heavily relying on the implementation)...

... so just to make sure, is this what you expect:

```python
@pytest.mark.parametrize("wrapper", ALL_COVER_FIXTURES, indirect=["wrapper"])
async def test_open(wrapper, hass, config):
    """Test cover opening."""

    feature_mock = wrapper.feature_mock
    def initial_update():
        feature_mock.state = 3  # manually stopped

    def open_gate():
        feature_mock.state = 1  # opening

    feature_mock.async_update = AsyncMock(side_effect=initial_update)
    feature_mock.async_open = AsyncMock(side_effect=open_gate)

    entity_id = (await wrapper.setup(hass, config)).entity_id
    assert hass.states.get(entity_id).state == STATE_CLOSED

    feature_mock.async_update = AsyncMock(side_effect=None)
    await hass.services.async_call(
        "cover",
        SERVICE_OPEN_COVER,
        {"entity_id": entity_id},
        blocking=True,
    )
    assert hass.states.get(entity_id).state == STATE_OPENING

```

?

MartinHjelmare replied:
I'm assuming that `feature_mock` is part of the library. And `async_setup_entity` sets up the integration. I think it looks good.

gadgetmobile commented:
I addressed all issues (in order) in commits 2d97e5ea8a..bd70c23e07 and rebased.

I'm finally beginning to like how the cover tests look - thanks for the help making them so much better!

MartinHjelmare commented on tests/components/blebox/test_config_flow.py:
```diff
@@ -0,0 +1,180 @@
+from tests.async_mock import DEFAULT, AsyncMock, PropertyMock, patch
+
+
+async def empty_config_flow_from_user(hass):
```
These helpers make it harder to see what the tests are doing. Please remove them.

gadgetmobile replied:
Done. (Though I personally would've at least refactored out the common `data=` parameter.)

gadgetmobile committed:
remove test helper methods

gadgetmobile committed:
assure that async_update is initially properly mocked

gadgetmobile committed:
remove wrapper class

gadgetmobile commented:
I think I've covered (no pun intended) all the suggestions - and I don't have any further ideas what to improve.

Feel free to be as picky as you like, so that I'll know how to properly prepare the next PR (after this one).

MartinHjelmare commented on tests/components/blebox/test_cover.py:
```diff
@@ -0,0 +1,380 @@
+    """Test gateController default state."""
+
+    feature_mock, entity_id = gatecontroller
+    entity = await async_setup_entity(hass, config, entity_id)
```
Please rename `entity` to `entry` or similar as it's an entity registry entry and not an entity instance.

MartinHjelmare commented on tests/components/blebox/test_init.py:
```diff
@@ -0,0 +1,63 @@
+        with pytest.raises(ConfigEntryNotReady):
+            config = mock_config()
+            config.add_to_hass(hass)
+            await async_setup_entry(hass, config)
```
Please use the core interface `hass.config_entries.async_setup` to set up the config entry.

MartinHjelmare commented on tests/components/blebox/test_init.py:
```diff
@@ -0,0 +1,63 @@
+    patch_product_identify(None, side_effect=blebox_uniapi.error.ClientError)
+
+    with patch("homeassistant.components.blebox._LOGGER.error") as error:
+        with pytest.raises(ConfigEntryNotReady):
```
Remove this and assert the state of the config entry instead.

MartinHjelmare commented on tests/components/blebox/test_init.py:
```diff
@@ -0,0 +1,63 @@
+
+    patch_product_identify(None, side_effect=blebox_uniapi.error.ClientError)
+
+    with patch("homeassistant.components.blebox._LOGGER.error") as error:
```
There's a `caplog` fixture available to assert log contents.

MartinHjelmare reviewed (commented):
Looks good! Some comments.

gadgetmobile committed:
rename entity to entry

gadgetmobile committed:
properly use async_setup and caplog in tests

gadgetmobile committed:
fix pylint warning about unused var

MartinHjelmare commented on tests/components/blebox/test_init.py:
```diff
@@ -0,0 +1,54 @@
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert "Identify failed at 172.100.123.4:80 ()" in caplog.text
```
We can also assert that the `entry.state` is `ENTRY_STATE_SETUP_RETRY` in this case.

MartinHjelmare commented on tests/components/blebox/test_init.py:
```diff
@@ -0,0 +1,54 @@
+    caplog.set_level(logging.ERROR)
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+    assert "Identify failed at 172.100.123.4:80 ()" in caplog.text
```
See above.

MartinHjelmare commented on tests/components/blebox/test_init.py:
```diff
@@ -0,0 +1,54 @@
+
+    await hass.config_entries.async_unload(entry.entry_id)
+    await hass.async_block_till_done()
+    assert not hass.data.get(DOMAIN)
```
Here the `entry.state` should be `ENTRY_STATE_NOT_LOADED`.

gadgetmobile committed:
check entry state properly

MartinHjelmare reviewed (approved):
Great!

bdraco reviewed (approved):
Nice cleanups since the last time I looked this over. @MartinHjelmare Thanks for jumping in.

gadgetmobile commented:
Big thanks to both of you for helping me out so much here!

bdraco commented:
Looks like one test left to fix
`FAILED tests/components/blebox/test_cover.py::test_stop[pyloop-gatebox] - Ass...`

gadgetmobile committed:
remove impossible test case

gadgetmobile commented:
That test started failing for me locally only after a rebase (likely related to #35073 changes).

To fix it, I removed that test case (since "cover stop service" wouldn't be available, because stop wouldn't be supported - so it was testing an impossible case).

I pushed this without the rebase for clarity. (And all tests now should pass with and without a rebase anyway).

I'm expecting CI to succeed now.
